src/utils.c

```diff
@@ -33,6 +33,40 @@ void rate_limitf(float *value, float target, float step) {
     }
 }
 
+// Smoothen changes in tilt angle by ramping the step size
+// smooth_center_window: Sets the angle away from Target that step size begins ramping down
+void smooth_rampf(
+    float *value,
+    float *ramped_step,
+    float target,
+    float step,
+    float smoothing_factor,
+    float smooth_center_window
+) {
+    float tiltback_target_diff = target - *value;
+
+    // Within X degrees of Target Angle, start ramping down step size
+    if (fabsf(tiltback_target_diff) < smooth_center_window) {
+        // Target step size is reduced the closer to center you are (needed for smoothly
+        // transitioning away from center)
+        *ramped_step = (smoothing_factor * step * (tiltback_target_diff / 2)) +
+            ((1 - smoothing_factor) * *ramped_step);
+        // Linearly ramped down step size is provided as minimum to prevent overshoot
+        float centering_step = fminf(fabsf(*ramped_step), fabsf(tiltback_target_diff / 2) * step) *
+            sign(tiltback_target_diff);
```

Comment:
I'm not really great at math, so I may totally be missing the point, but I find this computation overly complicated and ultimately flawed.

The `step` here is no longer the step in itself, it's a multiplier for the `tiltback_target_diff` and the step is thus a linear function of the diff. The way I understand it you are filtering towards the real step (`ramped_step`) if outside the window and you're filtering towards half of the real step when within the window.

The line above is what I think is not working really well, consider the case when you've just entered the `smooth_center_window` after you've filtered almost towards `step * tiltback_target_diff`, so that will be the value of `*ramped_step`. `centering_step`, though, after going through the `fminf()`, will be `fabsf(tiltback_target_diff / 2) * step`, which is almost half of that, so that transition won't be smooth at all. Or am I wrong on this?

I don't dispute it's working in a way (not sure why the above non-continuity doesn't noticeably manifest), but I want the math in the package to be solid. I think this is supposed to do in essence what @aeraglyx described in https://github.com/lukash/refloat/pull/2#issuecomment-2196212802, except his solution I think is (to the extent I can make that claim, again, I'm not really too good at math) mathematically correct. There are likely more ways of doing this too, e.g. I was thinking of maybe an exponential function (while this one is linear), etc.

Reply:
I have to admit that I simply copy pasted this code without questioning it from the Input-tilt implementation. Now that you forced me to look into it I have to say that something does look fishy about that code, good job for catching that.

Before further investigating I double-checked that I didn't introduce the problem during refactoring. Thus the problem should also affect the original input tilt implementation.

I think the problem starts when ramped_step_size is updated for the smooth center window. If the smooth center window is made large enough (>> 2), it appears as if ramped_step_size could actually temporarily increase instead of decreasing, since "tiltback_target_diff / 2" can be larger than 1.

I haven't had time yet to look into a fix but I figured I owed you a response acknowledging the problem you found.